providers/desec/protocol.go

```diff
@@ -0,0 +1,215 @@
+	Name       string      `json:"name,omitempty"`
+	Published  time.Time   `json:"published,omitempty"`
+}
+
+type resourceRecord struct {
+	Subname string   `json:"subname"`
+	Records []string `json:"records"`
+	TTL     uint32   `json:"ttl,omitempty"`
+	Type    string   `json:"type"`
+	Target  string   `json:"-"`
+}
+
+type rrResponse struct {
+	resourceRecord
+	Created time.Time `json:"created"`
+	Domain  string    `json:"domain"`
+	Name    string    `json:"name"`
+}
+
+type dnssecKey struct {
+	Dnskey  string   `json:"dnskey"`
+	Ds      []string `json:"ds"`
+	Flags   int      `json:"flags"`
+	Keytype string   `json:"keytype"`
+}
+
+type errorResponse struct {
+	Detail string `json:"detail"`
+}
+
+func (c *api) fetchDomainList() error {
+	c.domainIndex = map[string]uint32{}
+	var dr []domainObject
+	endpoint := "/domains/"
+	var bodyString, err = c.get(endpoint, "GET")
+	if err != nil {
+		return fmt.Errorf("Error fetching domain list from deSEC: %s", err)
+	}
+	err = json.Unmarshal(bodyString, &dr)
+	if err != nil {
+		return err
+	}
+	for _, domain := range dr {
+		//We store the min ttl in the domain index
+		//This will be used for validation and auto correction
+		c.domainIndex[domain.Name] = domain.MinimumTTL
+	}
+	return nil
+}
+
+func (c *api) getRecords(domain string) ([]resourceRecord, error) {
+	endpoint := "/domains/%s/rrsets/"
+	var rrs []rrResponse
+	var rrs_new []resourceRecord
+	var bodyString, err = c.get(fmt.Sprintf(endpoint, domain), "GET")
+	if err != nil {
+		return rrs_new, fmt.Errorf("Error fetching records from deSEC for domain %s: %s", domain, err)
+	}
+	err = json.Unmarshal(bodyString, &rrs)
+	if err != nil {
+		return rrs_new, err
+	}
+	// deSEC returns round robin records as array but dnsconfig expects single entries for each record
+	// we will create one object per record except of TXT records which are handled as array of string by dnscontrol aswell.
+	for i := range rrs {
+		tmp := resourceRecord{
+			TTL:     rrs[i].TTL,
+			Type:    rrs[i].Type,
+			Subname: rrs[i].Subname,
+			Records: rrs[i].Records,
+		}
+		rrs_new = append(rrs_new, tmp)
+	}
+	return rrs_new, nil
+}
+
+func (c *api) createDomain(domain string) error {
+	endpoint := "/domains/"
+	pl := domainObject{Name: domain}
+	byt, _ := json.Marshal(pl)
+	var resp []byte
+	var err error
+	if resp, err = c.post(endpoint, "POST", byt); err != nil {
+		return fmt.Errorf("Error create domain deSEC: %v", err)
+	}
+	dm := domainObject{}
+	err = json.Unmarshal(resp, &dm)
+	if err != nil {
+		return err
+	}
+	printer.Printf("If you want to use DNSSec please add the DS record at your registrar using one of the keys:\n")
+	printer.Printf("%+q", dm.Keys)
+	return nil
+}
+
+//upsertRR will create or override the RRSet with the provided resource record.
+func (c *api) upsertRR(rr []resourceRecord, domain string) error {
+	endpoint := fmt.Sprintf("/domains/%s/rrsets/", domain)
+	byt, _ := json.Marshal(rr)
+	if _, err := c.post(endpoint, "PATCH", byt); err != nil {
```

Comment:
Should this be `PATCH`?  I thought it should be `PUT` but I could be wrong. (Do they both work if the label doesn't previously exist?)

Reply:
Both work almost in the same way, PATCH allowed me to leave out the ttl for deletions. But its better to use PUT and define the full rrsets.